hax/hax/util.py

```diff
@@ -1819,6 +1828,65 @@ def cleanup_node_process_states(self):
         if not self.kv.kv_delete_in_transaction(keys):
             raise HAConsistencyException('KV deletion failed')
 
+    @repeat_if_fails()
+    def process_dynamic_fidk_lock(self) -> bool:
+        # Acquire lock to update last_updated_base_fidk.
+        # This will block until lock is acquired.
+        # Will break if any other exception than
+        # HAConsistencyException occurs.
+        try:
+            while not self.kv.kv_put('fidk_update_lock', 'true', cas=0):
+                sleep(1)
+            return True
+        except Exception:
+            return False
+
+    @repeat_if_fails()
+    def process_dynamic_fidk_unlock(self):
+        # Release fidk_update_lock.
+        # This will block until released.
+        try:
+            keys: List[KeyDelete] = [
+                KeyDelete(name='fidk_update_lock', recurse=True),
+            ]
+            while not self.kv.kv_delete_in_transaction(keys):
+                sleep(1)
+        except Exception:
+            raise RuntimeError('Unreachable')
+
+    @repeat_if_fails()
+    def get_process_next_dynamic_fidk_lock(self) -> int:
+        if self.process_dynamic_fidk_lock():
+            fidk = self.kv.kv_get('last_dynamic_fid_key/process',
+                                  recurse=False)
+            new_fidk = int(json.loads(fidk['Value'])) + 1
+            # Update dynamic fid key.
+            while not self.kv.kv_put('last_dynamic_fid_key/process',
+                                     json.dumps(str(new_fidk))):
+                sleep(1)
+            self.process_dynamic_fidk_unlock()
+        return new_fidk
+
+    @repeat_if_fails()
+    def alloc_next_process_fid(self, process_fid: Fid) -> Fid:
+        next_fidk = self.get_process_next_dynamic_fidk_lock()
+        fid_mask: Fid = ObjTMaskMap[ObjT.PROCESS]
+        fid_cont = process_fid.container
+        fid_key = process_fid.key + ((fid_mask.key * next_fidk) + next_fidk)
```

Comment:
@mssawant  Just a question: currently, we have assigned higher 32 bits as dynamic part. So later if we change the mask to lower 32, then this formula changes too, right? 
like in cfgen
`fidk = (((obj_mask.key ^ QW_F) * fidk + fidk) &
                obj_mask.key)`

Reply:
That's the idea, formula should not change.
Please refer to https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/802258945/High+Level+Design+of+Dynamic+Motr+FID+Allocation+for+Motr+Client+processes#Partial-Fid-allocation, here I have illustrated the cases with reserving high and low 32 bits both. The formula does not change. Let me know if you have any doubts.